Fix for bug #765998: reference count leaks in _mogrify().  Back-ported
from upstream changes in 2.4.3.


diff -Naur psycopg2-2.0.14.orig/psycopg/cursor_type.c psycopg2-2.0.14/psycopg/cursor_type.c
--- psycopg2-2.0.14.orig/psycopg/cursor_type.c	2010-02-28 14:29:43.000000000 -0500
+++ psycopg2-2.0.14/psycopg/cursor_type.c	2012-02-02 13:26:55.727661642 -0500
@@ -167,6 +167,7 @@
                         }
                         else {
                             /* no adapter found, raise a BIG exception */
+                            Py_DECREF(key);
                             Py_XDECREF(value);
                             Py_DECREF(n);
                             return -1;
@@ -174,14 +175,12 @@
                     }
 
                     Py_XDECREF(t); /* t dies here */
-                    /* after the DECREF value has the original refcnt plus 1
-                       if it was added to the dictionary directly; good */
-                    Py_XDECREF(value);
                 }
                 else {
                     /* we have an item with one extra refcnt here, zap! */
                     Py_DECREF(item);
                 }
+                Py_XDECREF(value);
                 Py_DECREF(key); /* key has the original refcnt now */
                 Dprintf("_mogrify: after value refcnt: "
                     FORMAT_CODE_PY_SSIZE_T,
